### ISSUE TITLE
Transform Space Constraints

### ISSUE THREAD

**Issue by marcusx2** [playcanvas/engine#3342]
Please see Unity's [constraints](https://docs.unity3d.com/Manual/Constraints.html) concept. [Zapworks Studio](https://docs.zap.works/studio/scripting/reference/node/functions/relativeto-setter/) also has this concept and it goes a step further because it allows you to constrain something to world space and also the screen, which is very useful. Imagine, for example, making a bullet go from the world to the screen. Or you have a soccer game and you make the ball's position relative to the current player that holds it. In augmented reality, this is used a lot to make a GUI that is in world space go to the screen if the tracked image is no longer in view for example.

Constraints is a very useful concept that I'd love to see implemented.

**Comment by Maksims** [comment]
This is already trivial by few lines of code.

**Comment by yaustar** [comment]
Agreed, this is currently possible with the current features of the engine.

**Comment by yaustar** [comment]
Re-opening as re-reading unity constraints, there are some features that may be worth writing utility scripts for 🤔 Wouldn't mind seeing a clear example of use though

**Comment by dev5spatial** [comment]
Yes, it's possible to do code without constraints. Constraints is just a concept to code in a particular way that makes sense in some situations. It's just nice to have. Whenever you want to think of an entity that is constrained to another in some way. Said entity can be constrained by multiple entities, and have a weight to define how much it is affect by other entities. And being able to animate this weight.

I did give some examples:
Imagine, for example, making a bullet go from the world to the screen. Or you have a soccer game and you make the ball's position relative to the current player that holds it. In a image tracked augmented reality, this is used a lot to make a GUI that is in world space go to the screen(with a nice transition animation) if the tracked image is no longer in view for example.

**Comment by marcusx2** [comment]
Yes, it's possible to do code without constraints. Constraints is just a concept to code in a particular way that makes sense in some situations. It's just nice to have. Whenever you want to think of an entity that is constrained to another in some way. Said entity can be constrained by multiple entities, and have a weight to define how much it is affect by other entities. And being able to animate this weight to dynamically change how much an entity is affected by other entities(in terms of position, rotation, scale, etc, depends on what you want to constrain).

I did give some examples:
Imagine, for example, making a bullet go from the world to the screen. Or you have a soccer game and you make the ball's position relative to the current player that holds it. In a image tracked augmented reality, this is used a lot to make a GUI that is in world space go to the screen(with a nice transition animation) if the tracked image is no longer in view for example.

**Comment by yaustar** [comment]
The football example is a bit too simple of an example as it's done quite easily by parenting and unparenting the football to the player, or the UI to and from the camera.

I'm quite intrigued by the weighting system, is there an example of a use case for that?

**Comment by marcusx2** [comment]
https://drive.google.com/file/d/1tYBs_grNBJ5Y4RBgsvvPEW4WW-Q9ma3D/view?usp=sharing

**Comment by Maksims** [comment]
This sounds like very specific use cases, with very specific logic implementation, which is all doable now with fairly simple code.

For a design of some system, we ideally want to answer on few things:
1. Is it enabling new things.
2. Is it only makes life easier by providing a vastly improved alternative.
3. Is it very niche or generic in realistic contexts.
4. Is it limiting and only covers part of related tasks, or is it generic and does not limit a lot.
5. Does it bloats API with some niche things (every kB matters), or it benefits the majority of engine users.
6. Does it saves a lot of engineering upfront cost of implementing similar systems by developers.

### Good examples of strong APIs:
1. Assets system and management to async load, know when things are added to registry or loaded, etc.
2. Sound system with instances of sounds, nodes for sound processing, various sound settings.
3. Generic tween system to interpolate data from one to another state, the control flow of tweens, easings, events, etc.

Doing anything described above without existing APIs, would require large time investment upfront to design a system. And would not be trivial to do with few lines of code without such APIs.

**Comment by marcusx2** [comment]
Like I said, I think constraints enable an alternative way of thinking and coding. Kinda like when you want to do something repetitive you can use recursion if it makes sense. That's the way I see it. In addition, you are also able to see the constraints in action in a timeline by animating the weight. So the constraint logic is previewable in realtime using timelines, which I think is another strong point.

I thought it would be a great addition, I use contraints very often in frameworks that allow it. Anyways, if you guys don't think it's worth the effort, that's ok. Oh well, bummer I guess.

**Comment by marcusx2** [comment]
Coming back here to give more input. I think one of the best advantages of constraints is when you couple it with real time preview via timeline by animating the weight. I can see in real time a UI in world space going to the screen (typical augmented reality scenario) without having to compile and run to check if it will work as expected.

Ultimately, I think one of the biggest strengths of game frameworks like playcanvas(and Unity, and Zapworks Studio) is the ability to preview gameplay in realtime as much as possible in the editor, to save time. Constraints, coupled with timelines, enhance this power. That's why Unity has it and Zapworks Studio has it, I think.

While you can do everything without constraints with code, I could do that using other frameworks like Phaser, etc. But with PlayCanvas constraint with timelines can be used to preview a significant amount of functionality in real time in the editor. It just make it so much easier, and to me that is one of the selling points of game frameworks with an editor like PlayCanvas, Unity and others.

**Comment by marcusx2** [comment]
@Maksims Please take a look at this [forum post](https://forum.playcanvas.com/t/how-can-i-make-something-go-to-the-screen/22283). You said it is trivial to do what I want with code, could you help me out? Thank you for your time.

**Comment by marcusx2** [comment]
Hey guys, just uploading a video showing the strength of the constraint system coupled with timeline support. It's the ability to code the game in the editor and preview movement before even getting to run the game.

https://user-images.githubusercontent.com/5656275/147484494-537584e5-113f-44e1-84a8-7499ae0e52c5.mp4

@leonidaspir

**Comment by mvaligursky** [comment]
Thanks for the suggestion and for providing the additional examples.

We think transform-space constraints would be better incubated as a separate library or add-on rather than added directly to the engine API at this stage. A complete constraint system introduces significant API and behavioral questions around multiple targets, weighting, coordinate spaces, evaluation order, cycles, and interaction with animation, scripts, and physics. Exploring those semantics independently would avoid committing the engine to an API before they have been validated in real-world use.

If a library matures and sees broader adoption, we can reconsider whether some or all of it belongs in the engine. Editor visualization and timeline support would also need to be considered separately.

Closing this issue for now.
